server/src/main/java/org/opensearch/wlm/stats/QueryGroupState.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.wlm.stats;
+
+import org.opensearch.search.ResourceType;
+
+import java.util.EnumMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * This class will keep the point in time view of the query group stats
+ */
+public class QueryGroupState {
+    /**
+     * completions at the query group level, this is a cumulative counter since the Opensearch start time
+     */
+    private final AtomicLong completions = new AtomicLong();
+
+    /**
+     * rejections at the query group level, this is a cumulative counter since the Opensearch start time
+     */
+    private final AtomicLong rejections = new AtomicLong();
+
+    /**
+     * this will track the cumulative failures in a query group
+     */
+    private final AtomicLong failures = new AtomicLong();
+
+    /**
+     * This will track total number of cancellations in the query group due to all resource type breaches
+     */
+    private final AtomicLong totalCancellations = new AtomicLong();
+
+    /**
+     * This is used to store the resource type state both for CPU and MEMORY
+     */
+    private final Map<ResourceType, ResourceTypeState> resourceState;
+
+    public QueryGroupState() {
+        resourceState = new EnumMap<>(ResourceType.class);
+        resourceState.put(ResourceType.CPU, new ResourceTypeState(ResourceType.CPU));
+        resourceState.put(ResourceType.MEMORY, new ResourceTypeState(ResourceType.MEMORY));
```

Comment:
I think this will be a good idea since the `ResourceType` is used generically across the codebase and might get changed. In those scenarios we will have un-necessary objects here. More controlled approach here can be as following
```
List<ResourceType> trackedResourceTypes = List.of(ResourceType.CPU, ResourceType.MEMORY);
....

for (ResourceType resourceType: trackedResourceTypes) {
       resourceState.put(resourceType, new ResourceTypeState(resourceType));
```

Reply:
Enabling this change in favor of the argument that this construct should be confined to wlm.